### DISCUSSION TITLE
Request object by key returning an optional

### DISCUSSION THREAD

**Discussion by nlohmann** [nlohmann/json#3501] [Ideas]
I saw the following snippet (see below) being used to avoid repeated code like

```cpp
if(j.contains("key")) { something(j["key"]) }
```

The idea is to add a getter function that returns `std::nullopt_t` in case an object key is not found. As this approach is very similar to the `value` function which returns a default value in this case, it may be a good addition to the public API.

* * *

```cpp
/// Optional element access
/// @returns empty if element does not exist
std::optional<json> optional_value(const std::string& name) const
{
    auto it = find(name);
    return it != end() ? *it : std::optional<Njson>();
}

/// Optional element access with value type
/// @returns empty if element doesn't exist or does not have the given type
template<typename ValueType>
std::optional<ValueType> optional_value(const std::string& name) const
{
    auto it = find(name);
    return (it != end() && it->type() == nlohmann::json(ValueType()).type())
        ? it->get<ValueType>()
        : std::optional<ValueType>();
}
```

**Comment by gregmarr** [comment]
I can see utility in these functions, but I'm not sure about the naming.  I'm not sure it's obvious that in the below snippet that `v1` is `json` and `v2` is `int`.

    auto v1 = j.optional_value("foo");
    auto v2 = j.optional_value<int>("foo");

The second is `value()` but returns an optional of the type, so that makes sense.  Isn't the first `get()` but returns an optional `json`?  So `optional_get`?

**Reply by nlohmann** [reply]
Yes - naming is important there. I am also not sure whether both functions should have the same name, because the second one is stricter and it may be surprising to not get a value even if the key is present (but the time is not matching).
